crypto/dlogproofs/partial_dlog_knowledge_ec.go

```diff
@@ -0,0 +1,160 @@
+		*types.ECGroupElement, *types.ECGroupElement,
+		*types.ECGroupElement, *types.ECGroupElement) {
+	prover.a1 = a1
+	prover.a2 = a2
+	prover.secret1 = secret1
+	r1 := common.GetRandomInt(prover.DLog.GetOrderOfSubgroup())
+	c2 := common.GetRandomInt(prover.DLog.GetOrderOfSubgroup())
+	z2 := common.GetRandomInt(prover.DLog.GetOrderOfSubgroup())
+	prover.r1 = r1
+	prover.c2 = c2
+	prover.z2 = z2
+	x1X, x1Y := prover.DLog.Exponentiate(a1.X, a1.Y, r1)
+	x2X, x2Y := prover.DLog.Exponentiate(a2.X, a2.Y, z2)
+	b2ToC2X, b2ToC2Y := prover.DLog.Exponentiate(b2.X, b2.Y, c2)
+	b2ToC2InvX, b2ToC2InvY := prover.DLog.Inverse(b2ToC2X, b2ToC2Y)
+	x2X, x2Y = prover.DLog.Multiply(x2X, x2Y, b2ToC2InvX, b2ToC2InvY)
+	
+	x1 := types.NewECGroupElement(x1X, x1Y)
+	x2 := types.NewECGroupElement(x2X, x2Y)
+	
+	// we need to make sure that the order does not reveal which secret we do know:
+	ord := common.GetRandomInt(big.NewInt(2))
+	
+	if ord.Cmp(big.NewInt(0)) == 0 {
+		prover.ord = 0
+		return x1, a1, b1, x2, a2, b2
+	} else {
+		prover.ord = 1
+		return x2, a2, b2, x1, a1, b1
+	}
+}
+
+func (prover *PartialECDLogProver) GetProofData(challenge *big.Int) (*big.Int, *big.Int,
+		*big.Int, *big.Int) {
+	c1 := new(big.Int).Xor(prover.c2, challenge)	
+
+	z1 := new(big.Int)
+	z1.Mul(c1, prover.secret1)
+	z1.Add(z1, prover.r1)
+	z1.Mod(z1, prover.DLog.GetOrderOfSubgroup())
+	
+	if prover.ord == 0 {
+		return c1, z1, prover.c2, prover.z2
+	} else {
+		return prover.c2, prover.z2, c1, z1
+	}
+}
+
+type PartialECDLogVerifier struct {
+	DLog              *dlog.ECDLog
+	x1                *types.ECGroupElement
+	x2                *types.ECGroupElement
+	a1                *types.ECGroupElement
+	b1                *types.ECGroupElement
+	a2                *types.ECGroupElement
+	b2                *types.ECGroupElement
+	challenge         *big.Int
+}
+
+func NewPartialECDLogVerifier(dlog *dlog.ECDLog) *PartialECDLogVerifier {
+	verifier := PartialECDLogVerifier{
+		DLog:         dlog,
+	}
+
+	return &verifier
+}
+
+func (verifier *PartialECDLogVerifier) SetProofRandomData(x1, a1, b1, x2, a2, 
+		b2 *types.ECGroupElement) {
+	verifier.x1 = x1
+	verifier.a1 = a1
+	verifier.b1 = b1
+	verifier.x2 = x2
+	verifier.a2 = a2
+	verifier.b2 = b2
+}
+
+func (verifier *PartialECDLogVerifier) GetChallenge() *big.Int {
+	challenge := common.GetRandomInt(verifier.DLog.GetOrderOfSubgroup())
+	verifier.challenge = challenge
+	return challenge
+}
+
+func (verifier *PartialECDLogVerifier) Verify(c1, z1, c2, z2 *big.Int) bool {
+	c := new(big.Int).Xor(c1, c2)	
+	if c.Cmp(verifier.challenge) == 0 {
+	} else {
+		return false	
+	}
+
+	left1, left2 := verifier.DLog.Exponentiate(verifier.a1.X, verifier.a1.Y, z1)
+	r1, r2 := verifier.DLog.Exponentiate(verifier.b1.X, verifier.b1.Y, c1)
+	right1, right2 := verifier.DLog.Multiply(r1, r2, verifier.x1.X, verifier.x1.Y)
+	if left1.Cmp(right1) == 0 && left2.Cmp(right2) == 0 {
+	} else {
+		return false
+	}
```

Comment:
Please put code in L144-150 in a separate function that returns `bool`. Then, call it twice inside this function. Since it seems to me that L152-159 are a duplicate of L144-150 with the exception that L156 below returns `true` while the first block just continues. So, use the result of the first call of the newly introduced function to determine whether to continue or return, and if we continue, return the result of the second call to this newly introduced function.

Also, inside this newly introduced function, please use this logic instead of the current one in L147-150 to:

```go
return left1.Cmp(right1) == 0 && left2.Cmp(right2) == 0
```